### PR TITLE
Graphql article

### PR DESCRIPTION
**Pull request by anupama-pathirage**

### CODE REVIEW

File: ballerina-graphql-with-mysql-datastore/bookstore/Ballerina.toml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+[package]
+distribution = "2201.0.2"
+
+[build-options]
+observabilityIncluded = true
+
+[[platform.java11.dependency]]
+groupId = "mysql"
+artifactId = "mysql-connector-java"
```
Comment (niveathika) [inline ballerina-graphql-with-mysql-datastore/bookstore/Ballerina.toml]: Cant we use the mysql driver import?

File: ballerina-graphql-with-mysql-datastore/bookstore/bookdb.bal [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+isolated function getBooks(string? title) returns Book[] {
+    Book[] books = [];
+    sql:ParameterizedQuery query = `SELECT b.title, b.published_year, a.name, a.country,
+                                    a.language FROM BOOK  b left  join AUTHOR a on b.author_id=a.author_id`;
+    if title is string {
+        query = `SELECT b.title, b.published_year, a.name, a.country, a.language FROM BOOK b left join 
+                 AUTHOR a on b.author_id=a.author_id where  b.title=${title}`;
+    }
```
Comment (niveathika) [inline ballerina-graphql-with-mysql-datastore/bookstore/bookdb.bal]:
```suggestion
    if title is string {
             query = sql: queryConcat(query, `where b.title=${title}`);
          }
```

File: ballerina-graphql-with-mysql-datastore/bookstore/bookstore.bal [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        return getBooks(());
+    }
+
+    remote function addBook(string authorName, string authorCountry, string authorLanguage, string title,
+                                                                        int published_year) returns Book? {
+        int|error ret = addBookData(authorName, authorCountry, authorLanguage, title, published_year);
+        if ret is error {
+            return ();
+        }
+        return getBooks(title)[0];
```
Comment (DimuthuMadushan) [inline ballerina-graphql-with-mysql-datastore/bookstore/bookstore.bal]:
```suggestion
        if ret is int {
            return getBooks(title)[0];
        }
        return;
```

File: ballerina-graphql-with-mysql-datastore/bookstore/bookdb.bal [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                title: <string>result["title"],
+                published_year: <int>result["published_year"],
+                author: {
+                    name: <string>result["name"],
+                    country: <string>result["country"],
+                    language: <string>result["language"]
+                }
+            };
+            books.push(b);
+        };
```
Comment (niveathika) [inline ballerina-graphql-with-mysql-datastore/bookstore/bookdb.bal]:
```suggestion
    type Data record {|
    int book_id;
    string title;
    int author_id;
    string name;
    string country;
|};

    stream<Data, error?> resultStream = dbClient->query(`SELECT b.book_id, b.title, a.author_id, a.name, a.country 
                                                                FROM BOOK  b left  join AUTHOR a on b.author_id=a.author_id`);
    Book[]? books = check from var {book_id, title, author_id, name, country} in resultStream
        select {
            book_id: book_id,
            title: title,
            author: {
                author_id: author_id,
                name: name,
                country: country
            }
        };
    return books == () ? [] : books;
```

File: ballerina-graphql-with-mysql-datastore/bookstore/bookdb.bal [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    if res is error {
+        log:printError(res.message());
+    }
+    return books;
+}
+
+function addBookData(string authorName, string authorCountry, string authorLanguage, string title, int published_year)
+                                                                                                    returns error|int {
+    int author_id = -1;
+    stream<record {}, error?> resultStream = dbClient->query(`SELECT author_id from AUTHOR WHERE name=${authorName}`);
```
Comment (niveathika) [inline ballerina-graphql-with-mysql-datastore/bookstore/bookdb.bal]: Shall we use queryRow() method for this? This way if NoRowsError is thrown, we can add the author else add book.

File: ballerina-graphql-with-mysql-datastore/bookstore/bookstore.bal [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+service /bookstore on new graphql:Listener(4000) {
+
+    isolated resource function get bookByName(string title) returns Book[] {
+        return getBooks(title);
+    }
+
+    resource function get allBooks() returns Book[] {
+        return getBooks(());
+    }
```
Comment (DimuthuMadushan) [inline ballerina-graphql-with-mysql-datastore/bookstore/bookstore.bal]:
```suggestion
    isolated resource function get allBooks() returns Book[] {
        return getBooks(());
    }
```

File: ballerina-graphql-with-mysql-datastore/bookstore/bookstore.bal (hunk 3 above)
Comment (anupama-pathirage) [reply ballerina-graphql-with-mysql-datastore/bookstore/bookstore.bal]: Changed the logic. Thanks.

File: ballerina-graphql-with-mysql-datastore/bookstore/Ballerina.toml (hunk 1 above)
Comment (anupama-pathirage) [reply ballerina-graphql-with-mysql-datastore/bookstore/Ballerina.toml]: Done. Thanks.

File: ballerina-graphql-with-mysql-datastore/bookstore/bookdb.bal (hunk 2 above)
Comment (anupama-pathirage) [reply ballerina-graphql-with-mysql-datastore/bookstore/bookdb.bal]: Done. Thanks.

File: ballerina-graphql-with-mysql-datastore/bookstore/bookdb.bal (hunk 4 above)
Comment (anupama-pathirage) [reply ballerina-graphql-with-mysql-datastore/bookstore/bookdb.bal]: Done. Thanks.

File: ballerina-graphql-with-mysql-datastore/bookstore/bookstore.bal (hunk 6 above)
Comment (anupama-pathirage) [reply ballerina-graphql-with-mysql-datastore/bookstore/bookstore.bal]: Done. Thanks.

**Pull request merged**